app.py

```diff
@@ -23,68 +50,97 @@ def modify_query(**new_values):
+    nextPage = ""
 
+    if len(results[page*2*8:(page*2*8)+8]) == 0 or len(results[page*8:(page*8)+8]) < 4:
+        nextPage = ""
+    elif len(results) > 0:
+        nextPage = "search?key=" + key + "&value=" + value + "&page=" + str(page + 1)
+
+    if page == 0:
+        lastPage = ""
+    elif len(results) > 0:
+        lastPage = "search?key=" + key + "&value=" + value + "&page=" + str(page - 1)
+
+    return render_template('search.j2', results=results[page*8:(page*8)+8], lastPage=lastPage, nextPage=nextPage) 
 
 @app.route("/app")
-def application():
-    package = OpenShopChannel.package_by_name(request.args.get("package"))
-    if package:
-        return render_template('app.html', package=package)
+def appPage():
+    selectedApp = request.args.get('app', default = 'danbo', type = str)
+    selectedApp = OpenShopChannel.package_by_name(selectedApp)
+    
+    readableSize = sizeof_fmt(selectedApp["extracted"])
+
+    return render_template('app.j2', app=selectedApp, size=readableSize)
+
+@app.route("/random")
+def randomApp():
+    response = Response('')
+    response.headers["Cache-Control"] = "no-cache, no-store, must-revalidate"
+    response.headers["Location"] = "/app?app=" + random.choice(OpenShopChannel.get_packages())["internal_name"]
+    return response, 301
+
+@app.route("/error")
+def errorPage():
+    errorcode = request.args.get('error', default = 'danbo', type = str)
+    errortext = getErrorText(errorcode)
+    if errorcode == "SUCCESS":
+        return redirect("/", code=302)
     else:
-        return render_template('error.html', code="APP-404",
-                               description="The app \"{}\" (Internal Name) could not be found.".format(
-                                   request.args.get("package")))
+        return render_template('error.j2', errorcode=errorcode, errortext=errortext)
 
-
-# Errors
 @app.errorhandler(404)
 def page_not_found(e):
-    return render_template('error.html', code="HTTP-404",
-                           description="Page could not be found. This is not where the shop is :(")
+    errorcode = "HTTP_404"
+    errortext = "The requested page could not be found."
+    return render_template('error.j2', errorcode=errorcode, errortext=errortext)
 
 
 @app.errorhandler(500)
 def server_error(e):
-    return render_template('error.html', code="HTTP-500",
-                           description="Something went wrong on the shop. Please report this event!")
+    errorcode = "HTTP_500"
+    errortext = "The server has encountered an error. This isn't your fault- try your action again."
+    return render_template('error.j2', errorcode=errorcode, errortext=errortext)
 
 
 if __name__ == '__main__':
     WSGIRequestHandler.protocol_version = "HTTP/1.1"
-    context = ssl.SSLContext(ssl.PROTOCOL_TLSv1)
-    context.load_cert_chain('cert.pem', 'key.pem')
-
-    app.run(host='127.0.0.1', port=443, debug=True, ssl_context=context)
+    app.run(host='0.0.0.0', port=80, debug=True)
```

Comment:
No, that was just for testing, thanks for catching that. I would be happy to set that back up.  
  
That has got me thinking, though- implementing SSL would require patching our root certificate into the channel, which is quite a few more bytes than is necessary for functionality witth plain HTTP. It is absolutely possible, though, and people like @spotlightishere have already done so, considering the working key + cert you sent me earlier.  
  
Considering secure information is not sent between the client and the server, however, I'm not yet aware of the advantages it would give us. With two bytes patched, the shop channel will trust all domains & HTTP connections for JavaScript & Opera purposes:

![](https://user-images.githubusercontent.com/16677360/132048044-2a0961d0-43bc-4697-98e9-35221a2045d0.png)

This is, ultimately, up to everyone, and if I am missing an advantage of HTTPS, or if HTTPS is just a preference for privacy purposes, whatever works works!

Reply:
@spotlightishere @SketchMaster2001 @dakrk